Mains/Tracks_test.cc

```diff
@@ -455,44 +487,142 @@ int main(int argc, char **argv) {
+//          mctraj.append(endtraj);
+//        }
+        tracker.simulateHits(*trkfield,mctraj,hits,xings,trackerinters,trckrIWinters,trckrECDWSinters,mctol);
         tinfo_.ncells = xings.size();
         tinfo_.ntrkhits = hits.size();
         tinfo_.narcs = trackerinters.size();
         if(hits.size() >= minnhits){
 // calcluate the estart energy loss
+
+          if (trckrIWinters.size()>0) {
+            int iwintotfst=0;
+            if (ntrkrECUPS_<1) { iwintotfst=1; }
+            ntrkrIWotr_ = trckrIWinters.size()-iwintotfst;
+            for (unsigned int iwint=iwintotfst; iwint<trckrIWinters.size(); ++iwint) {
+              auto iwlinter = trckrIWinters.at(iwint);
+              trkrIWotrde_ += mctraj.energy(iwlinter.end())-mctraj.energy(iwlinter.begin());
+              auto posIn = mctraj.position3(iwlinter.begin());
+              auto posOut = mctraj.position3(iwlinter.end());
+              if ( iwint == (trckrIWinters.size()-1) ) {
+                trckZout = posIn.Z();
+              }
+              trkrIWotrdpath_ += (posOut-posIn).R();
+           }
+          }
+          if (trckrECDWSinters.size()>0) {
+            auto ewlinter = trckrIWinters.at(0);
+            ntrkrECDWS_ = trckrECDWSinters.size();
+            trkrECDWSe_ = mctraj.energy(ewlinter.end());
+            trkrECDWSde_ = trkrECDWSe_ - mctraj.energy(ewlinter.begin());
+            auto posIn = mctraj.position3(ewlinter.begin());
+            auto posOut = mctraj.position3(ewlinter.end());
+            trckZout = posIn.Z();
+            trkrECDWSpath_ += (posOut-posIn).R();
+          }
+
           double trackerpath(0.0);
           for(auto const& inter : trackerinters) { trackerpath += speed*inter.range(); }
 //          double ke = cestate.energy() - cestate.mass();
 //          trackerde_ = -100*trackerEStar.dEIonization(ke)*tracker.density()*trackerpath; // unit conversion
-          trackere_ = mctraj.energy(mctraj.range().end());
-          trackerde_ = trackere_ - ipae_;
+//          trackere_ = mctraj.energy(mctraj.range().end());
+          trackere_ = mctraj.energy(trackerinters.back().end());
+          trackerde_ = trackere_ - bfrtrcke_;//ipae_;
           tarde->Fill(targetde_);
           ipade->Fill(ipade_);
           trkde->Fill(trackerde_);
           trknc->Fill(tinfo_.ncells);
+          ipaEqde->Fill(ipade_+trkrWentrde_);
           // add calo hit
           tinfo_.ncalohits = calo.simulateHits(*trkfield,mctraj,hits,mctol);
           // truncate the true trajectory
           mctraj.setRange(TimeRange(mctraj.range().begin(),mctraj.back().range().begin()+0.1));
           // get the true times at entrance and exit
-          double mctent = ztime(mctraj,trackerinters.front().begin(),tracker.zMin());
-          double mctext = ztime(mctraj,trackerinters.back().end(),tracker.zMax());
+          double mctent = ztime(mctraj,trackerinters.front().begin(),trckZin/*tracker.zMin()*/);
```

Comment:
Can you explain why the z limits need to be overwritten?  Couldn't you instead of modified tracker.zMin() to take into account the walls (if they are non-null)?